common/flatpak-run.c

```diff
@@ -2441,35 +2445,61 @@ forward_file (XdpDbusDocuments *documents,
 
   fd_list = g_unix_fd_list_new ();
   fd_id = g_unix_fd_list_append (fd_list, fd, error);
+  if (fstat (fd, &stbuf) == 0 && S_ISDIR (stbuf.st_mode))
+    is_dir = TRUE;
   close (fd);
 
-  if (!xdp_dbus_documents_call_add_sync (documents,
-                                         g_variant_new ("h", fd_id),
-                                         TRUE, /* reuse */
-                                         FALSE, /* not persistent */
-                                         fd_list,
-                                         &doc_id,
-                                         NULL,
-                                         NULL,
-                                         error))
-    {
-      if (error)
-        g_dbus_error_strip_remote_error (*error);
-      return FALSE;
-    }
+  flags = DOCUMENT_ADD_FLAGS_REUSE_EXISTING;
+  if (is_dir)
+    flags |= DOCUMENT_ADD_FLAGS_DIRECTORY;
+
+  /* Requires version 4 of AddFull interface when exporting directory */
+  if (!xdp_dbus_documents_call_add_full_sync (documents,
+                                              g_variant_new_fixed_array (G_VARIANT_TYPE_HANDLE, &fd_id, 1, sizeof (gint32)),
+                                              flags,
+                                              app_id,
+                                              perms,
+                                              fd_list,
+                                              &doc_ids,
+                                              NULL,
+                                              NULL,
+                                              NULL,
+                                              error))
+    {
+      /* Fallback to plain org.freedesktop.portal.Documents.Add and
```

Comment:
So we really don't want to fall back in all cases, but only if it is not supported. Unfortunately it seems like the dbus error doesn't really tell us that we're trying to do something not supported in this version. So I think we should check the proxies `version` property against the required version (which changes depending on the flags).